src/test/spec/retryable_reads.rs

```diff
@@ -0,0 +1,176 @@
+
+            let db_name = match test_file.database_name {
+                Some(ref db_name) => db_name.clone(),
+                None => test_case.description.replace('$', "%").replace(' ', "_"),
+            };
+
+            let coll_name = match test_file.collection_name {
+                Some(ref coll_name) => coll_name.clone(),
+                None => "coll".to_string(),
+            };
+
+            let coll = client.init_db_and_coll(&db_name, &coll_name).await;
+
+            if let Some(ref data) = test_file.data {
+                match data {
+                    TestData::Single(data) => {
+                        if !data.is_empty() {
+                            coll.insert_many(data.clone(), None)
+                                .await
+                                .expect(&test_case.description);
+                        }
+                    }
+                    TestData::Many(_) => panic!("{}: invalid data format", &test_case.description),
+                }
+            }
+
+            if let Some(ref fail_point) = test_case.fail_point {
+                client
+                    .database("admin")
+                    .run_command(fail_point.clone(), None)
+                    .await
+                    .unwrap();
+            }
+
+            let mut events: Vec<TestEvent> = Vec::new();
+            for operation in test_case.operations {
+                let result = match operation.object {
+                    OperationObject::Client => client.run_client_operation(&operation).await,
+                    OperationObject::Database => {
+                        client.run_database_operation(&operation, &db_name).await
+                    }
+                    OperationObject::Collection => {
+                        client
+                            .run_collection_operation(&operation, &db_name, &coll_name)
+                            .await
+                    }
+                    OperationObject::GridfsBucket => {
+                        panic!("unsupported operation: {}", operation.name)
+                    }
+                };
+                let mut operation_events: Vec<TestEvent> = client
+                    .collect_events(&operation, false)
+                    .into_iter()
+                    .map(Into::into)
+                    .collect();
+
+                if let Some(error) = operation.error {
+                    assert_eq!(
+                        result.is_err(),
+                        error,
+                        "{}: expected error: {}, got {:?}",
+                        test_case.description,
+                        error,
+                        result
+                    );
+                }
+                if let Some(expected_result) = operation.result {
+                    let expected_result = match expected_result {
+                        Bson::Int32(n) => Bson::Int64(n as i64),
+                        _ => expected_result,
+                    };
```

Comment:
[This update](https://github.com/mongodb/bson-rust/blob/cbcd60c850476cb6f1c17f221cc0be9f331f7249/src/extjson/de.rs#L188) broke the tests for operations that return an `i64` since the expected value in the json tests is always relatively small (i.e. less than `i32::MAX`), and was therefore being deserialized into a `Bson::Int32` and then compared against a `Bson::Int64`. This workaround just converts any expected result that's a `Bson::Int32` to a `Bson::Int64`, and should be fine to do in any case since there aren't any methods in the driver that return an `i32`.

Reply:
I think the `match` module that Patrick wrote should be able to handle this: https://github.com/mongodb/mongo-rust-driver/blob/c4562c28d2e6b6a9330352fadb60e938ad6f5439/src/test/util/matchable.rs